src/renderer/opengl.rs

```diff
@@ -16,63 +17,202 @@ use glutin::{
 };
 use glutin_winit::DisplayBuilder;
 use raw_window_handle::HasRawWindowHandle;
-use winit::dpi::PhysicalSize;
+use skia_safe::{
+    canvas::Canvas,
+    gpu::{
+        backend_render_targets::make_gl, gl::FramebufferInfo, surfaces::wrap_backend_render_target,
+        DirectContext, SurfaceOrigin,
+    },
+    ColorType,
+};
 use winit::{
-    event_loop::EventLoop,
+    dpi::PhysicalSize,
+    event_loop::{EventLoop, EventLoopProxy},
     window::{Window, WindowBuilder},
 };
 
-pub struct GlWindow {
-    pub window: Window,
-    config: Config,
-}
+#[cfg(target_os = "windows")]
+pub use super::vsync::VSyncWinDwm;
+
+#[cfg(target_os = "macos")]
+pub use super::vsync::VSyncMacos;
+
+use super::{SkiaRenderer, VSync, WindowConfig, WindowConfigType};
 
-pub struct Context {
-    surface: Surface<WindowSurface>,
+use crate::{profiling::tracy_gpu_zone, window::UserEvent};
+
+#[cfg(feature = "gpu_profiling")]
+use crate::profiling::{opengl::create_opengl_gpu_context, GpuCtx};
+
+pub struct OpenGLSkiaRenderer {
+    // NOTE: The destruction order is important, so don't re-arrange
+    // If possible keep it the reverse of the initialization order
+    skia_surface: skia_safe::Surface,
+    fb_info: FramebufferInfo,
+    pub gr_context: DirectContext,
     context: PossiblyCurrentContext,
-    window: Window,
+    window_surface: Surface<WindowSurface>,
     config: Config,
+    window: Window,
 }
 
-pub fn clamp_render_buffer_size(size: PhysicalSize<u32>) -> PhysicalSize<u32> {
+fn clamp_render_buffer_size(size: &PhysicalSize<u32>) -> PhysicalSize<u32> {
     PhysicalSize::new(
         size.width.clamp(1, MAX_RENDERBUFFER_SIZE),
         size.height.clamp(1, MAX_RENDERBUFFER_SIZE),
     )
 }
 
-impl Context {
-    pub fn window(&self) -> &Window {
-        &self.window
-    }
+fn get_proc_address(surface: &Surface<WindowSurface>, addr: &CStr) -> *const c_void {
+    GlDisplay::get_proc_address(&surface.display(), addr)
+}
+
+impl OpenGLSkiaRenderer {
+    pub fn new(window: WindowConfig, srgb: bool, vsync: bool) -> Self {
+        #[allow(irrefutable_let_patterns)] // This can only be something else than OpenGL on Windows
+        let config = if let WindowConfigType::OpenGL(config) = window.config {
+            config
+        } else {
+            panic!("Not an opengl window");
+        };
+        let window = window.window;
+        let gl_display = config.display();
+        let raw_window_handle = window.raw_window_handle();
+
+        let size = clamp_render_buffer_size(&window.inner_size());
+
+        let surface_attributes = SurfaceAttributesBuilder::<WindowSurface>::new()
+            .with_srgb(Some(srgb))
+            .build(
+                raw_window_handle,
+                NonZeroU32::new(size.width).unwrap(),
+                NonZeroU32::new(size.height).unwrap(),
```

Comment:
Those unwraps might very well blow up if starting in e.g. an initially minimized state, need to try that out.

Reply:
The code is just moved, so it has been the same before. `clamp_render_buffer_size` should ensure that it's not zero.